packages/@aws-cdk/aws-ec2/lib/launch-template.ts

```diff
@@ -0,0 +1,688 @@
+   * You can use a duration of 1, 2, 3, 4, 5, or 6 hours.
+   *
+   * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances
+   *
+   * @default Requested spot instances do not have a pre-defined duration.
+   */
+  readonly blockDuration?: Duration;
+
+  /**
+   * The behavior when a Spot Instance is interrupted.
+   *
+   * @default Spot instances will terminate when interrupted.
+   */
+  readonly interruptionBehavior?: SpotInstanceInterruption;
+
+  /**
+   * Maximum hourly price you're willing to pay for each Spot instance. The value is given
+   * in dollars. ex: 0.01 for 1 cent per hour, or 0.001 for one-tenth of a cent per hour.
+   *
+   * @default Maximum hourly price will default to the on-demand price for the instance type.
+   */
+  readonly maxPrice?: number;
+
+  /**
+   * The Spot Instance request type.
+   *
+   * If you are using Spot Instances with an Auto Scaling group, use one-time requests, as the
+   * Amazon EC2 Auto Scaling service handles requesting new Spot Instances whenever the group is
+   * below its desired capacity.
+   *
+   * @default One-time spot request.
+   */
+  readonly requestType?: SpotRequestType;
+
+  /**
+   * The end date of the request. For a one-time request, the request remains active until all instances
+   * launch, the request is canceled, or this date is reached. If the request is persistent, it remains
+   * active until it is canceled or this date and time is reached.
+   *
+   * @default The default end date is 7 days from the current date.
+   */
+  readonly validUntil?: Expiration;
+};
+
+/**
+ * The types of resource tags that can be set on a launch template.
+ */
+export enum LaunchTemplateTagResourceType {
+  // dev-note: Full list at https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-tagspecification.html#cfn-ec2-launchtemplate-tagspecification-resourcetype
+  // These are the only ones from the list that presently support tag on create.
+
+  /**
+   * EC2 Instances.
+   */
+  INSTANCE = 'instance',
+
+  /**
+   * EBS Volumes.
+   */
+  VOLUME = 'volume',
+}
+
+/**
+ * Interface for defining the tag specification of a LaunchTemplate.
+ */
+export interface LaunchTemplateTagSpecification {
+  /**
+   * The types of resources that are tagged with the given tags. Note that there are
+   * restrictions regarding which resources can be tagged when created. See the following
+   * for additional information.
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-tagspecification.html#cfn-ec2-launchtemplate-tagspecification-resourcetype
+   */
+  readonly resourceTypes: LaunchTemplateTagResourceType[];
+
+  /**
+   * The tags to apply to the resources.
+   */
+  readonly tags: CfnTag[];
+}
+
+/**
+ * Properties of a LaunchTemplate.
+ */
+export interface LaunchTemplateProps {
```

Comment:
I defaulted with making all of the properties optional, since they all are completely optional in a Launch Template. Oddly, it is valid to create a completely empty Launch Template.

My thinking is that the PR is a good place to have a discussion regarding how opinionated the CDK's L2 should be regarding which properties are optional and which are not.

Four of the CloudFormation resources that can make use of LaunchTemplates are:
1. [EC2::EC2Fleet](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-ec2fleet.html)  -- Uses an array of them, where each element of the array can [override some of the LaunchTemplate values](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-ec2fleet-fleetlaunchtemplateconfigrequest.html#cfn-ec2-ec2fleet-fleetlaunchtemplateconfigrequest-overrides).
2. [EC2::SpotFleet](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-spotfleet.html) -- Similar to EC2::Fleet in that it can use an array of LaunchTemplates, but the values that can be overriden [are different](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-spotfleet-launchtemplateconfig.html#cfn-ec2-spotfleet-launchtemplateconfig-overrides).
3. [EC2::Instance](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html) -- Can use a single launch template, and the properties (ex: machine image) defined by the Launch Template are overriden by defining the equivalent top-level property in the EC2::Instance resource.
4. [AutoScaling::AutoScalingGroup](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html) -- Can use either [a single Launch Template with no overrides](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html#cfn-as-group-launchtemplate). Or a [mixed instances policy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-mixedinstancespolicy.html) which uses a single launch template as a base template, but may [optionally provide an override launch template](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-launchtemplateoverrides.html) for each element of an array of instance types.

Reply:
This is one of the reasons we ourselves have been getting stuck implementing it in the CDK. It just seems so open-ended and anything can be overridden at any point.

Thanks for this list! This is super helpful.